lang/ir/src/ir/attrs.rs

```diff
@@ -0,0 +1,979 @@
+                use crate::error::ExtError as _;
+                return Err(format_err!(
+                    attr.span(),
+                    "encountered duplicate ink! attribute"
+                )
+                .into_combine(format_err!(seen.span(), "first ink! attribute here")))
+            }
+            seen.insert(attr);
+        }
+        Ok(())
+    }
+}
+
+impl TryFrom<syn::Attribute> for Attribute {
+    type Error = syn::Error;
+
+    fn try_from(attr: syn::Attribute) -> Result<Self, Self::Error> {
+        if attr.path.is_ident("ink") {
+            return <InkAttribute as TryFrom<_>>::try_from(attr).map(Into::into)
+        }
+        Ok(Attribute::Other(attr))
+    }
+}
+
+impl From<InkAttribute> for Attribute {
+    fn from(ink_attribute: InkAttribute) -> Self {
+        Attribute::Ink(ink_attribute)
+    }
+}
+
+impl TryFrom<syn::Attribute> for InkAttribute {
+    type Error = syn::Error;
+
+    fn try_from(attr: syn::Attribute) -> Result<Self, Self::Error> {
+        if !attr.path.is_ident("ink") {
+            return Err(format_err_spanned!(attr, "unexpected non-ink! attribute"))
+        }
+        match attr.parse_meta().map_err(|_| {
+            format_err_spanned!(attr, "unexpected ink! attribute structure")
+        })? {
+            syn::Meta::List(meta_list) => {
+                let args = meta_list
+                    .nested
+                    .into_iter()
+                    .map(<AttributeArg as TryFrom<_>>::try_from)
+                    .collect::<Result<Vec<_>, syn::Error>>()?;
+                Self::ensure_no_duplicate_flags(&args)?;
+                if args.is_empty() {
+                    return Err(format_err_spanned!(
+                        attr,
+                        "encountered unsupported empty ink! attribute"
+                    ))
+                }
+                Ok(InkAttribute { args })
+            }
+            _ => Err(format_err_spanned!(attr, "unknown ink! attribute")),
+        }
+    }
+}
+
+impl InkAttribute {
+    /// Returns `Ok` if the given iterator yields no duplicate attribute flags.
+    ///
+    /// # Errors
+    ///
+    /// If the given iterator yields duplicate attribute flags.
+    fn ensure_no_duplicate_flags<'a, I>(args: I) -> Result<(), syn::Error>
+    where
+        I: IntoIterator<Item = &'a AttributeArg>,
+    {
+        use crate::error::ExtError as _;
+        use std::collections::HashSet;
+        let mut seen: HashSet<&AttributeArg> = HashSet::new();
+        for arg in args.into_iter() {
+            if let Some(seen) = seen.get(arg) {
+                return Err(format_err!(
+                    arg.span(),
+                    "encountered duplicate ink! attribute arguments"
+                )
+                .into_combine(format_err!(
+                    seen.span(),
+                    "first equal ink! attribute argument here"
+                )))
+            }
+            seen.insert(arg);
+        }
+        Ok(())
```

Comment:
Duplicate code with `ensure_no_duplicate_args` (and maybe `ensure_no_duplicate_attrs`)

Reply:
Thanks! Eliminated the duplicate. However, while similar, `ensure_no_duplicate_attrs` actually does some slightly different thing.